src/os_version_check/darwin_impl.rs

```diff
@@ -0,0 +1,590 @@
+///
+/// The size of the parts are inherently limited by Mach-O's `LC_BUILD_VERSION`.
+#[inline]
+pub(super) const fn pack_os_version(major: u16, minor: u8, patch: u8) -> OSVersion {
+    let (major, minor, patch) = (major as u32, minor as u32, patch as u32);
+    (major << 16) | (minor << 8) | patch
+}
+
+/// We'd usually use `CStr::from_bytes_until_nul`, but that can't be used in `compiler-builtins`.
+#[inline]
+fn trim_trailing_nul(mut bytes: &[u8]) -> &[u8] {
+    while let Some((b'\0', rest)) = bytes.split_last() {
+        bytes = rest;
+    }
+    bytes
+}
+
+/// Parse an OS version from a bytestring like b"10.1" or b"14.3.7".
+#[track_caller]
+pub(super) const fn parse_os_version(bytes: &[u8]) -> OSVersion {
+    let (major, bytes) = parse_usize(bytes);
+    if major > u16::MAX as usize {
+        panic!("major version is too large");
+    }
+    let major = major as u16;
+
+    let bytes = if let Some((period, bytes)) = bytes.split_first() {
+        if *period != b'.' {
+            panic!("expected period between major and minor version")
+        }
+        bytes
+    } else {
+        return pack_os_version(major, 0, 0);
+    };
+
+    let (minor, bytes) = parse_usize(bytes);
+    if minor > u8::MAX as usize {
+        panic!("minor version is too large");
+    }
+    let minor = minor as u8;
+
+    let bytes = if let Some((period, bytes)) = bytes.split_first() {
+        if *period != b'.' {
+            panic!("expected period after minor version")
+        }
+        bytes
+    } else {
+        return pack_os_version(major, minor, 0);
+    };
+
+    let (patch, bytes) = parse_usize(bytes);
+    if patch > u8::MAX as usize {
+        panic!("patch version is too large");
+    }
+    let patch = patch as u8;
+
+    if !bytes.is_empty() {
+        panic!("too many parts to version");
+    }
+
+    pack_os_version(major, minor, patch)
+}
+
+#[track_caller]
+const fn parse_usize(mut bytes: &[u8]) -> (usize, &[u8]) {
+    // Ensure we have at least one digit (that is not just a period).
+    let mut ret: usize = if let Some((&ascii, rest)) = bytes.split_first() {
+        bytes = rest;
+
+        match ascii {
+            b'0'..=b'9' => (ascii - b'0') as usize,
+            _ => panic!("found invalid digit when parsing version"),
+        }
+    } else {
+        panic!("found empty version number part")
+    };
+
+    // Parse the remaining digits.
+    while let Some((&ascii, rest)) = bytes.split_first() {
+        let digit = match ascii {
+            b'0'..=b'9' => ascii - b'0',
+            _ => break,
+        };
+
+        bytes = rest;
+
+        // This handles leading zeroes as well.
+        match ret.checked_mul(10) {
+            Some(val) => match val.checked_add(digit as _) {
+                Some(val) => ret = val,
+                None => panic!("version is too large"),
+            },
+            None => panic!("version is too large"),
+        };
+    }
+
+    (ret, bytes)
+}
```

Comment:
Unfamiliar with `compiler-builtins`, so I can't tell if there's a better way of doing this kind of parsing?

Reply:
Things are helped by this being specific to Apple platforms, so we already depend on `libSystem.dylib`, and could use for example `sscanf`. What do you prefer?

Reply:
Actually, I think I'd prefer that myself, it's also what LLVM does (the manual parsing here was a remnant of [my implementation in `objc2`](https://github.com/madsmtm/objc2/blob/53a0cb58649a9ca1951d39544bef1cbaa7210d90/crates/objc2/src/__macro_helpers/os_version/apple.rs)). Will wait 'till I hear back, then I'll probably implement that.

Reply:
I'll take a closer look at the parsing in a bit, but it is probably better not to depend on `sscanf` for the reason mentioned in https://github.com/rust-lang/compiler-builtins/pull/794#discussion_r2000295614.